controllers/storagecluster/storageconsumer.go

```diff
@@ -190,24 +190,39 @@ func (s *storageConsumer) ensureDeleted(r *StorageClusterReconciler, storageClus
 	storageConsumer := &ocsv1a1.StorageConsumer{}
 	storageConsumer.Name = defaults.LocalStorageConsumerName
 	storageConsumer.Namespace = storageCluster.Namespace
-	if err := r.Get(r.ctx, client.ObjectKeyFromObject(storageConsumer), storageConsumer); err != nil {
-		if client.IgnoreNotFound(err) != nil {
-			return ctrl.Result{}, fmt.Errorf("failed to get storageconsumer %s: %v", storageConsumer.Name, err)
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(storageConsumer), storageConsumer); client.IgnoreNotFound(err) != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get storageconsumer %s: %v", storageConsumer.Name, err)
+	} else if storageConsumer.UID != "" {
+		if storageConsumer.Status.Client != nil {
+			return ctrl.Result{}, fmt.Errorf("waiting for client to offboard before deleting storageconsumer %s", storageConsumer.Name)
 		}
-		return ctrl.Result{}, nil
-	}
 
-	if err := r.Delete(r.ctx, storageConsumer); err != nil {
-		return ctrl.Result{}, fmt.Errorf("failed to delete storageconsumer %s: %v", storageConsumer.Name, err)
-	}
+		if err := r.Delete(r.ctx, storageConsumer); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to delete storageconsumer %s: %v", storageConsumer.Name, err)
+		}
 
-	if controllerutil.RemoveFinalizer(storageConsumer, internalComponentFinalizer) {
-		r.Log.Info("Removing finalizer from StorageConsumer.", "StorageConsumer:", storageConsumer.Name, " StorageConsumer Namespace:", storageConsumer.Namespace, " Finalizer:", internalComponentFinalizer)
-		if err := r.Update(r.ctx, storageConsumer); err != nil {
-			r.Log.Info("Failed to remove finalizer from StorageConsumer.", "StorageConsumer:", storageConsumer.Name, "Finalizer:", internalComponentFinalizer)
-			return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from StorageConsumer: %v", err)
+		if controllerutil.RemoveFinalizer(storageConsumer, internalComponentFinalizer) {
```

Comment:
Shouldn't the finalizer removal block be before the Delete call?

Reply:
This is the internal component finalizer that is added from storagecluster to the storageconsumer we can remove it after the deletion call